File: 1923-sentence-similarity-iii/sentence-similarity-iii.cpp
class Solution {
public:
    bool areSentencesSimilar(string sentence1, string sentence2) {
        std::deque<std::string> deque1, deque2;
        std::istringstream stream1(sentence1), stream2(sentence2);
        std::string word;

        while (stream1 >> word) {
            deque1.push_back(word);
        }
        while (stream2 >> word) {
            deque2.push_back(word);
        }

        while (!deque1.empty() && !deque2.empty() &&
               deque1.front() == deque2.front()) {
            deque1.pop_front();
            deque2.pop_front();
        }

        while (!deque1.empty() && !deque2.empty() &&
               deque1.back() == deque2.back()) {
            deque1.pop_back();
            deque2.pop_back();
        }

        return deque1.empty() || deque2.empty();
    }
};

File: 1923-sentence-similarity-iii/sentence-similarity-iii.java
class Solution {
    public boolean areSentencesSimilar(String sentence1, String sentence2) {
        String first = sentence1;
        String second = sentence2;
        if (first.length() == second.length()) {
            return first.equals(second);
        }
        if (sentence2.length() < first.length()) {
            first = sentence2;
            second = sentence1;
        }
        int i = -1;
        while (i + 1 < first.length() && first.charAt(i + 1) == second.charAt(i + 1)) {
            i++;
        }
        int j = first.length();
        int last = second.length();
        while (j - 1 >= 0 && first.charAt(j - 1) == second.charAt(last - 1)) {
            j--;
            last--;
        }

        if (i == first.length() - 1 && second.charAt(i + 1) == ' ') {
            return true;
        }
        if (j == 0 && second.charAt(last - 1) == ' ') {
            return true;
        }
        if (i + 1 >= j && second.charAt(i) == ' ' && second.charAt(last) == ' ') {
            return true;
        }

        return false;
    }
}